[WIP] TBB + backend refactor... DO NOT MERGE!

stuartarchibald opened the pull request:
This refactors the parallel backend and adds TBB.

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented:
I've just thrown this into the buildfarm to see what happens:
* windows 64: build fail, visual studio 9.0 cannot even parse `tbbpool.cpp` and also complains about exception handlers in use but no unwind semantics enabled.
* windows 32: build fail, cannot compile `omppool.cpp`, `alloca` identifier not found and `r` index in OMP for must be signed integral type.
* OSX: build fail, cannot find `"<omp.h>"` for `omppool.cpp`.
* linux 32 and linux 64: build passes, segfaults during testing somewhere in TBB.
* POWER8: passes, but slightly suspicious this is due to tests not actually running. Need to augment the `numba -s` tool with threading layer detection.
* ARMv8, build passes, tests also pass, but same suspicion as for POWER8.

* Python 2.7: tests fail as there's no `multiprocessing.get_context`.

stuartarchibald commented:
Status as of 8276a62 

| Arch | OMP build | TBB build | Testing | Notes|
| :---: | --- | --- | --- | --- |
| ARM|FAIL (Missing symbol `GOMP_parallel`) | FAIL (no TBB pkg) | N/A ||
| OSX Python 2| FAIL (no `<omp.h>`) | PASS | N/A  ||
| OSX Python 3| FAIL (no `<omp.h>`) | PASS| N/A  ||
| ppc64le Python 2| PASS | FAIL (no TBB pkg) | Hangs |GNU OpenMP forksafe issue?|
| ppc64le Python 3| PASS | FAIL (no TBB pkg) | Hangs |GNU OpenMP forksafe issue?|
| Windows 32 Python 2|N/A| FAIL | N/A||
| Windows 32 Python 3|PASS| PASS | Scipy missing https://github.com/numba/numba/pull/3257, fork ctx not available on windows|
| Windows 64 Python 2|N/A| FAIL | N/A||
| Windows 64 Python 3|PASS| PASS | Scipy missing https://github.com/numba/numba/pull/3257, fork ctx not available on windows|
| Linux 32 Python 2|PASS |PASS | Segfaults | GNU OpenMP forksafe issue?|
| Linux 32 Python 3|PASS |PASS | Segfaults | GNU OpenMP forksafe issue?|
| Linux 64 Python 2|PASS |PASS | Segfaults | GNU OpenMP forksafe issue?|
| Linux 64 Python 3|PASS |PASS | Segfaults | GNU OpenMP forksafe issue?|

stuartarchibald commented:
Status as of 1c71742 

| Arch | OMP build | TBB build | Testing | Notes|
| :---: | --- | --- | --- | --- |
| ARM|FAIL (Missing symbol `GOMP_parallel`) | FAIL (no TBB pkg) | N/A ||
| OSX Python 2| PASS | PASS | PASS  ||
| OSX Python 3|PASS | PASS| FAIL  | libomp/libiomp5 collide. Also, `@jit` is having some issues in the function cache|
| ppc64le Python 2| PASS | FAIL (no TBB pkg) | PASS |OpenMP testing only|
| ppc64le Python 3| PASS | FAIL (no TBB pkg) | PASS |OpenMP testing only|
| Windows 32 Python 2|N/A| FAIL | N/A|TBB pool C++ advanced for compiler|
| Windows 32 Python 3|PASS| PASS | PASS|Sporadic failure from hang/cash, not sure if it's related|
| Windows 64 Python 2|N/A| FAIL | N/A|TBB pool C++ advanced for compiler|
| Windows 64 Python 3|PASS| PASS |Sporadic failure from hang/cash, not sure if it's related|
| Linux 32 Python 2|PASS |PASS | Segfaults |OpenMP pool sometimes fails to import, TBB segfaults|
| Linux 32 Python 3|PASS |PASS | Segfaults |OpenMP pool has missing symbol `GOMP_parallel`, TBB segfaults|
| Linux 64 Python 2|PASS |PASS | Segfaults |OpenMP pool sometimes fails to import, TBB segfaults|
| Linux 64 Python 3|PASS |PASS | Segfaults |OpenMP pool has missing symbol `GOMP_parallel`, TBB segfaults|

anton-malakhov commented:
I've found out the reason for segfaults. In short, I think that anaconda's tbb package is broken because GCC 7 possibly over-optimizes TBB code:

Thread 7 received signal SIGSEGV, Segmentation fault.
tbb::internal::generic_scheduler::allocate_task (this=this@entry=0x7faddd75fe00, number_of_bytes=number_of_bytes@entry=8, parent=parent@entry=0x0,
    context=context@entry=0x7fade6008c60 <tbb::internal::the_dummy_context>) at ../../src/tbb/scheduler.cpp:320
320                 ITT_NOTIFY( sync_acquired, &my_return_list );
(rr) bt
`#0`  tbb::internal::generic_scheduler::allocate_task (this=this@entry=0x7faddd75fe00, number_of_bytes=number_of_bytes@entry=8, parent=parent@entry=0x0,
    context=context@entry=0x7fade6008c60 <tbb::internal::the_dummy_context>) at [../../src/tbb/scheduler.cpp:320](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/scheduler.cpp#L320)
`#1`  0x00007fade5deef2e in tbb::internal::generic_scheduler::generic_scheduler (this=this@entry=0x7faddd75fe00, m=...) at [../../src/tbb/scheduler.cpp:100](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/scheduler.cpp#L100)
`#2`  0x00007fade5df10c8 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::custom_scheduler (m=..., this=0x7faddd75fe00) at [../../src/tbb/custom_scheduler.h:59](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/custom_scheduler.h#L59)
`#3`  tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::allocate_scheduler (m=...) at [../../src/tbb/custom_scheduler.h:120](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/custom_scheduler.h#L120)

(rr) p t
**$7 = (tbb::task *) 0xffffffffffffffff**     This is the reason for segfault

(rr) f 3
#3  tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::allocate_scheduler (m=...) at ../../src/tbb/custom_scheduler.h:120
120             scheduler_type* s = new( p ) scheduler_type( m );
(rr) disassemble
Dump of assembler code for function tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::allocate_scheduler(tbb::internal::market&):
   0x00007fade5df10a0 <+0>:     push   %rbp
   0x00007fade5df10a1 <+1>:     push   %rbx
   0x00007fade5df10a2 <+2>:     xor    %edx,%edx
   0x00007fade5df10a4 <+4>:     mov    %rdi,%rbp
   0x00007fade5df10a7 <+7>:     mov    $0x188,%esi
   0x00007fade5df10ac <+12>:    mov    $0x1,%edi
   0x00007fade5df10b1 <+17>:    sub    $0x8,%rsp
   0x00007fade5df10b5 <+21>:    callq  0x7fade5dda8c8   # [call NFS_Alloc](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/custom_scheduler.h#L118)
   0x00007fade5df10ba <+26>:    mov    %rbp,%rsi
   0x00007fade5df10bd <+29>:    mov    %rax,%rdi
   0x00007fade5df10c0 <+32>:    mov    %rax,%rbx
   0x00007fade5df10c3 <+35>:    callq  [0x7fade5deeeb0 <tbb::internal::generic_scheduler::generic_scheduler(tbb::internal::market&)>](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/custom_scheduler.h#L120)
=> 0x00007fade5df10c8 <+40>:    lea    0x213579(%rip),%rax        # 0x7fade6004648 <_ZTVN3tbb8internal16custom_schedulerINS0_20IntelSchedulerTraitsEEE>

So, there are calls in lines 118 and 120 but call to std::memset on [line 119 ](https://github.com/01org/tbb/blob/tbb_2018/src/tbb/custom_scheduler.h#L119) was optimized out! Thus the `my_return_list`  member variable was never initialized (after re-allocating previously used chunk of memory) and thus allocate_task which is part of the generic_scheduler initialization sequence finds garbage and segfaults.

stuartarchibald commented:
Thanks for looking at the TBB segfault @anton-malakhov, seems like that would explain the problem.

I think we'll need to patch TBB? [Notes](https://en.cppreference.com/w/cpp/string/byte/memset) under `std::memset` suggest using e.g. `C11`'s `memset_s`.

stuartarchibald commented:
RE fork-exec, yes, noted, thanks. I spoke with @sklam about this mid-last week as the test suite itself also has a parallel testing flag (`-m`), which uses multiprocessing to distribute the test workload across cores and in practice is what is always used to make the tests run in a reasonable time.

`<speculation>`
This `-m` flag on linux + Python 2.7 will cause multiprocessing to end up using `fork(3)` with no preventative measure available. Under Python 3 the multiprocessing context may be set to use `spawn` for the test running processes but it is (`libc`/kernel!) implementation detail as to whether this is achieved via `fork(3)+exec(3)`.

In the following `OpenMP` refers to GNU OpenMP unless stated otherwise.

Under Python3, `multiprocessing.Process` from `spawn` via `fork(3)+exec(3)`:
For the tests we have on this branch I think they may be "safe" if no OpenMP calls are made in the parent thread hierarchy (assuming no `libgomp` DSO globals), this preventing any OpenMP threads spawning and any OpenMP mutexes being present. Then, repeated sequential fork+exec are potentially ok as the new process will contain the threadpool/mutexes for OpenMP which will then disappear with the child as it joins/exits.

Under Python2, `multiprocessing.Process`:
I cannot see how using OpenMP can be made safe unless `os.spawn*` is used to force `fork(3)+exec(3)` (assuming the C library implements it this way, GLIBC seems to https://sourceware.org/git/gitweb.cgi?p=glibc.git;a=blob;f=sysdeps/posix/spawni.c;h=b138ab4393cabbd8d1bec663bb8a000288ed1680;hb=HEAD), in which case this is back to the Python3 case.

The above however does not cover the unit tests that use the threading backend but are not protected by this mechanism. Should the above be valid, this is ok for internal R&D/testing code but probably not something that users should be forced to deal with. I think perhaps we could add an (`#ifdef`'d for GNU OpenMP) `__pthread_atfork()` registered function to the `omppool.c` that prints a message and aborts or similar to prevent accidental use, this can be added in the `launch_threads` part of the API (also note that `omp_set_num_threads` needs moving out of there as that would be a parent call to an OpenMP API)? I don't think it'd be possible to park the GNU OpenMP threads in a safe manner prior to fork as per TBB or Intel OpenMP as there's no API for it (a proposed patch is here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60035, mailing list discussion https://gcc.gnu.org/ml/gcc-patches/2014-02/msg00813.html).
`</speculation>`

anton-malakhov commented:
The answer about TBB was that it looks like `-flifetime-dse=1` is missed (probably by using CXXFLAGS instead, supplied by conda-build).

After replacing tbb package from intel channel, all the tests pass on linux, including all the fork-exec parts, so there should no be problem for intel runtimes I believe. The GNU OpenMP conclusions are sad.. but unless we contribute those API or create our own patch for Anaconda, there is nothing we can do here besides documenting or/and adding asserts/warnings.

seibert commented:
OK, we can open issues to get the TBB builds on conda-forge and Anaconda fixed as well.

stuartarchibald commented:
The GNU OpenMP conclusions are indeed a little sad. It seems in practice that `multiprocessing` `spawn` context does `fork()+exec()` in a consistent, detectable and GNU OpenMP safe manner. So I've added some code to try and detect situations which are unsafe (`multiprocessing` `fork` context) and handle them via raising `SIGTERM` in the child and printing to `stderr`. I've also found that Numba does not behave well under `os.fork()` either but that's out of scope of something we should try and rectify here, there's already enough to do!

anton-malakhov commented on docs/source/reference/envvars.rst:
```diff
@@ -278,3 +278,15 @@ Threading Control
+   threading library. Other valid values are:
+
+   * ``tbbpool`` - A threading layer backed by Intel TBB.
+   * ``omppool`` - A threading layer backed by OpenMP.
```
I suggest make it to respond to just `tbb`/`TBB` and `omp`/`OMP` like in the threading selection logic. However, in the testing, it relies on the full name

stuartarchibald replied:
Good idea, we can fix that up.

anton-malakhov commented on setup.py:
```diff
@@ -135,27 +137,77 @@ def get_ext_modules():
+        # Python 2 on windows has a VS dep which cannot handle sufficiently
+        # advanced C++ syntax to compile tbbpool.cpp.
+        tbb_unsupported = sys.platform.startswith('win') and \
+                          sys.version_info < (3, 0)
```
could this be specified in recipe instead? E.g. if no tbb is present, no `TBBROOT` is specified, then it will not be compiled anyway, no need for additional check. Otherwise, it breaks Intel's build (unless patched) since we use icc 13 which can compile it for py2win

stuartarchibald replied:
Yes. I think that is possible and essentially the same effort, good point about ICC.

anton-malakhov commented on buildscripts/condarecipe.local/meta.yaml:
```diff
@@ -27,6 +28,7 @@ requirements:
     - llvmlite 0.25.*
     - funcsigs                 # [py27]
     - singledispatch           # [py27]
+    - tbb-devel                # [not (armv6l or armv7l or aarch64)]
```
TBB should actually work there, is it just not ready on anaconda channel?

stuartarchibald replied:
I think so, fundamentally it's just C++ so there's a reasonable chance of it compiling fine, just hasn't been tried yet.

anton-malakhov commented on buildscripts/incremental/test.sh:
```diff
@@ -21,6 +21,10 @@ export NUMBA_DEVELOPER_MODE=1
 export PYTHONFAULTHANDLER=1
 
+if [ "$TEST_TBB" == "yes" ]; then
+export NUMBA_THREADING_LAYER='tbbpool'
```
can we change it to `NUMBA_THREADING_LAYER=TBB` so that it is closer to how users would like to see it (and they might be learning from this line as well)

stuartarchibald replied:
I think this needs adjustment anyway as it was in early on before e.g. omp was added, happy to accommodate this.

anton-malakhov commented on docs/source/reference/envvars.rst:
```diff
@@ -278,3 +278,16 @@ Threading Control
+   This environment variable controls the library used for concurrent execution
+   for the CPU parallel targets (``@vectorize(target='parallel')``, 
+   ``@guvectorize(target='parallel')``  and ``@njit(parallel=True)``). The 
+   variable type is string and by default is ``workqueue``, the Numba internal
```
You discuss `default` below, so probably it's out of sync here

stuartarchibald replied:
Thanks, not updated this since the new criteria based backend selection was added. Will change.

anton-malakhov commented on docs/source/user/threading-layer.rst:
```diff
@@ -0,0 +1,119 @@
+
+* ``tbb`` - A threading layer backed by Intel TBB.
+* ``omp`` - A threading layer backed by OpenMP.
+* ``workqueue`` - A threading layer backed by OS native threads.
```
well.. it's not like TBB and OMP are using some magic non-native threads ;-) Probably, `Simple work-sharing task scheduler`

stuartarchibald replied:
I was hoping to emphasise that there is no magic, it's just Numba using OS native threads with no 3rd party library involved, "built-in" does this fine, will change.

anton-malakhov commented on numba/npyufunc/parallel.py:
```diff
@@ -35,58 +38,48 @@ def get_thread_count():
+    domain partitioning.
 
+    NOTE: The execution backend is passed the requested thread count, but it can
+    choose to ignore it (TBB)!
```
let's probably just remove it? the only affected implementation is workqueue, and there is a way to store the chosen number of threads

stuartarchibald commented:
@anton-malakhov Just FYI: I've cherry-picked @sklam's PR patches from https://github.com/numba/numba/pull/3278 onto this branch to help with the problems caused by concurrent access in the finalizer.

stuartarchibald commented:
I think this may be the best option, else there's going to be continuous conflicts present at the DSO level especially when e.g. NumPy et al. are involved. Do you recall if the `intel-openmp` package provides the `omp.h` headers etc?

anton-malakhov commented:
no, there are no headers, but it is not a big problem since it is compatible with native openmp for respective platforms and comply with common OpenMP specification (and we don't want to use Intel's extensions so far). So, it's a matter of proper linking - during build time or in run-time.

stuartarchibald commented:
Great, thanks for the info, I think I can make this work through `llvm-openmp` for header provision and then `intel-openmp` for the linkage, it'll make for an interesting build configuration!

stuartarchibald commented:
Status as of a23bc19  

| Arch | OMP build | TBB build | Testing | Notes|
| :---: | --- | --- | --- | --- |
| ARM Python 3|PASS | N/A | PASS |No TBB package available|
| OSX Python 2| PASS | PASS | Sporadic fail  | Now using `libiomp5`. Failure due to pickle of closure in Numba's linalg impl|
| OSX Python 3|  PASS | PASS | PASS  | Now using `libiomp5`|
| ppc64le Python 2| PASS | PASS | PASS ||
| ppc64le Python 3| PASS | PASS | PASS ||
| Windows 32 Python 2|PASS| N/A | PASS| TBB disabled (won't compile)|
| Windows 32 Python 3|PASS| PASS | PASS||
| Windows 64 Python 2|PASS| N/A | PASS|TBB disabled (won't compile)|
| Windows 64 Python 3|PASS| PASS |PASS||
| Linux 32 Python 2|PASS |PASS | Sporadic fail | Failure due to pickle of closure in Numba's linalg impl|
| Linux 32 Python 3|PASS |PASS | PASS | |
| Linux 64 Python 2|PASS |PASS | Sporadic fail | Failure due to pickle of closure in Numba's linalg impl|
| Linux 64 Python 3|PASS |PASS | PASS | |

stuartarchibald commented:
Pulled in this https://github.com/numba/numba/pull/3291 to get accurate CPU information.

stuartarchibald commented:
0932314 passed CI, it's just codecov that failed because I masked off all the tests apart from those impacted by this change to get faster turn around on build state, 31eb370 reinstates all the tests.

stuartarchibald commented on rebld.sh:
```diff
@@ -0,0 +1,3 @@
+rm build/*/numba/npyufunc/tbbpool.* build/*/numba/npyufunc/*.so numba/npyufunc/workqueue.*.so
```
Reminder to delete this.

stuartarchibald replied:
done

stuartarchibald commented:
@anton-malakhov Since merging in your PR to switch off debug / the last build pass, I've:
 * switched off Python debugging
 * removed the thread count arg from the `parallel_for` signature
 * Added back in some missing asserts in testing
 * Added testing for `numba.threading_layer()` [currently failing on windows... investigating now]
 * Updated the documentation with regards to requirements for the different backends, and also defined the practice of directly setting a backend as "advanced usage".
 * Run autoformatters over the code.

Am going to move this to ready-for-review once I've sorted windows out.

stuartarchibald commented:
Build farm smoke test built the conda packages ok for all platforms on py2/3.
